src/gridfs/download.rs

```diff
@@ -139,16 +142,210 @@ impl GridFsBucket {
             n += 1;
         }
 
-        let expected_n =
-            total_bytes_expected / chunk_size + u64::from(total_bytes_expected % chunk_size != 0);
-        if (n as u64) != expected_n {
+        if n != file.n() {
             return Err(ErrorKind::GridFs(GridFsErrorKind::WrongNumberOfChunks {
                 actual_number: n,
-                expected_number: expected_n as u32,
+                expected_number: file.n() as u32,
             })
             .into());
         }
 
         Ok(())
     }
 }
+
+pub struct GridFsDownloadStream {
+    state: State,
+    current_n: u32,
+    file: FilesCollectionDocument,
+}
+
+type GetBytesFuture = BoxFuture<'static, Result<(Vec<u8>, Cursor<Chunk<'static>>)>>;
+
+enum State {
+    // Idle stores these fields as options so that they can be moved into a GetBytesFuture
+    // without requiring ownership of the state. They will always store values and can be
+    // unwrapped safely.
+    Idle {
+        buffer: Option<Vec<u8>>,
+        cursor: Box<Option<Cursor<Chunk<'static>>>>,
+    },
+    Busy(GetBytesFuture),
+    Done,
+}
+
+impl State {
+    fn set_busy(&mut self, new_future: GetBytesFuture) -> &mut GetBytesFuture {
+        *self = State::Busy(new_future);
+        match self {
+            Self::Busy(ref mut future) => future,
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl GridFsDownloadStream {
+    async fn new(
+        file: FilesCollectionDocument,
+        chunks: &Collection<Chunk<'static>>,
+    ) -> Result<Self> {
+        let initial_state = if file.length == 0 {
+            State::Done
+        } else {
+            let options = FindOptions::builder().sort(doc! { "n": 1 }).build();
+            let cursor = chunks.find(doc! { "files_id": &file.id }, options).await?;
+            State::Idle {
+                buffer: Some(Vec::new()),
+                cursor: Box::new(Some(cursor)),
+            }
+        };
+        Ok(Self {
+            state: initial_state,
+            current_n: 0,
+            file,
+        })
+    }
+
+    /// Gets the file `id` for the stream.
+    pub fn files_id(&self) -> &Bson {
+        &self.file.id
+    }
+}
+
+impl AsyncRead for GridFsDownloadStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::result::Result<usize, futures_util::io::Error>> {
+        let stream = self.get_mut();
+
+        let future = match &mut stream.state {
+            State::Idle { buffer, cursor } => {
+                let mut taken_buffer = buffer.take().unwrap();
+                let taken_cursor = cursor.take().unwrap();
+
+                if !taken_buffer.is_empty() {
+                    let bytes_to_write = std::cmp::min(taken_buffer.len(), buf.len());
+                    buf[..bytes_to_write]
+                        .copy_from_slice(taken_buffer.drain(..bytes_to_write).as_slice());
+                    if taken_cursor.has_next() {
```

Comment:
Think we also need to verify that the buffer is empty here before transitioning to done (e.g. if the input buffer is less than half the chunk size). This would actually be a good test case to cover I think.

Reply:
Good catch, this is fixed by the refactor you suggested above. I updated the test to include this scenario.